### ISSUE TITLE
JsonElement.ArrayEnumerator breaks when using Lazy<> / Nullable<> wrapper

### ISSUE THREAD

**Issue by mhehle** [dotnet/runtime#89384]
### Description

`JsonElement.ArrayEnumerator` cannot be used in combination with `Nullable` or `Lazy` wrappers.
When using the `Current` property an `InvalidOperationException` "Operation is not valid due to the current state of the object." occurs. The underlying `JsonElement` loses the reference to the parent `JsonDocument`.

#### Unit Tests (nunit)

```c#
public class JsonElementArrayEnumeratorTest
{
    private JsonDocument _jsonDocument;

    private string json = """
    {
        "array": [
            {
                "field": "value1"
            },
            {
                "field": "value2"
            }
        ]
    }
    """;

    [SetUp]
    public void SetUp()
    {
        _jsonDocument = JsonDocument.Parse(json);
    }

    [Test]
    public void ArrayEnumerator_WithNoWrapper()
    {
        var arrayElement = _jsonDocument.RootElement.GetProperty("array");
        var enumerator = arrayElement.EnumerateArray();
        
        // go to first array element
        enumerator.MoveNext();

        // first element should be an object
        Assert.That(enumerator.Current.ValueKind, Is.EqualTo(JsonValueKind.Object));
        Assert.That(enumerator.Current.GetProperty("field").GetString(), Is.EqualTo("value1"));
    }

    [Test]
    public void ArrayEnumerator_WithLazyWrapper()
    {
        var arrayElement = _jsonDocument.RootElement.GetProperty("array");
        // get lazy array enumerator
        var enumerator = new Lazy<JsonElement.ArrayEnumerator>(() => arrayElement.EnumerateArray());

        // go to first array element
        enumerator.Value.MoveNext();

        // first element should be an object
        Assert.That(enumerator.Value.Current.ValueKind, Is.EqualTo(JsonValueKind.Object));
        Assert.That(enumerator.Value.Current.GetProperty("field").GetString(), Is.EqualTo("value1"));
    }
    
    [Test]
    public void ArrayEnumerator_WithNullableWrapper()
    {
        var arrayElement = _jsonDocument.RootElement.GetProperty("array");
        JsonElement.ArrayEnumerator? enumerator = arrayElement.EnumerateArray();

        // go to first array element
        enumerator.Value.MoveNext();

        // first element should be an object
        Assert.That(enumerator.Value.Current.ValueKind, Is.EqualTo(JsonValueKind.Object));
        Assert.That(enumerator.Value.Current.GetProperty("field").GetString(), Is.EqualTo("value1"));
    }
}
```

### Expected behavior

`JsonElement.ArrayEnumerator` should work with `Nullable` or `Lazy` wrappers

### Actual behavior

When using the `Current` property an `InvalidOperationException` "Operation is not valid due to the current state of the object." occurs. The underlying `JsonElement` loses the reference to the parent `JsonDocument`.

### Regression?

No

### Known Workarounds

-

#### uname -e
```bash
6.1.38-1-MANJARO
```

#### dotnet --info
```bash
.NET SDK:
 Version:   7.0.306
 Commit:    f500069cb7

Laufzeitumgebung:
 OS Name:     manjaro
 OS Version:  
 OS Platform: Linux
 RID:         manjaro-x64
 Base Path:   /usr/share/dotnet/sdk/7.0.306/

Host:
  Version:      7.0.9
  Architecture: x64
  Commit:       8e9a17b221

.NET SDKs installed:
  6.0.403 [/usr/share/dotnet/sdk]
  6.0.408 [/usr/share/dotnet/sdk]
  6.0.412 [/usr/share/dotnet/sdk]
  7.0.101 [/usr/share/dotnet/sdk]
  7.0.203 [/usr/share/dotnet/sdk]
  7.0.306 [/usr/share/dotnet/sdk]
```

**Comment by eiriktsarpalis** [comment]
This is not related to either `Nullable` or `Lazy` per se, but the fact that you are using these types to wrap mutable structs. Every time you run `container.Value.MoveNext()` you are actually running `MoveNext()` on a copy of the contained value and not the contained value itself. For example I can reproduce the same error in the following example:
```C#
using System.Text.Json;

var arrayElement = JsonDocument.Parse("[1]").RootElement;
var enumerator = arrayElement.EnumerateArray();

var enumeratorCopy = enumerator;
enumeratorCopy.MoveNext();

Console.WriteLine(enumerator.Current.ValueKind); // Undefined
Console.WriteLine(enumerator.Current.GetInt32()); // InvalidOperationException
```

TL;DR avoid wrapping mutable structs in Lazy or Nullable. Consider either wrapping reference types (e.g. the `JsonDocument` instance itself) or only using immutable structs only.
